File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int l=0,r=0,maxlen=0,maxf=0;
        int[] hash = new int[26];
        while(r<s.length()){
            int val = s.charAt(r) - 'A';
            hash[val] += 1;
            maxf = Math.max(maxf,hash[val]);
            if(r-l+1 - maxf > k){
                hash[s.charAt(l) - 'A']--;
                l = l+1;
            }
            if(r-l+1 - maxf <= k){
                maxlen = Math.max(maxlen, r-l+1);
            }
            r++;
        }
        return maxlen;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        start = 0
        fmap = {}
        maxf = 0
        longest = 0

        for end in range(len(s)):
            fmap[s[end]] = fmap.get(s[end], 0) + 1

            maxf = max(maxf, fmap[s[end]])

            is_val = (end + 1 - start - maxf <= k)

            if not is_val:
                fmap[s[start]] -= 1
                start += 1
            
            longest = end + 1 - start
    
        return longest